### PR TITLE
End-to-end automation test for reindexing #1751

### PR DESCRIPTION
**Pull request by prb112**
This PR addresses: 

1 - Hardening the Reindex Operation to bad input and reflective responses
2 - Update the documentation to include an omitted input parameter 
3 - Add Reindex workflow with three jobs (in the matrix) - db2, postgres, derby
4 - Add multiple phase tests which require restarts between them
5 - BasicDate.json had a small error using valueString instead of code
6 - Added LogicalResource and Type level for the ReindexOperation
7 - Cleaned up the FHIROperationRegistry to register at system, type and instance level

What's not included: 
1 - Reindex after Migration (a separate PR, based on these integration tests, will address the tests)

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: build/reindex/bin/gather-logs.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+###############################################################################
+# (C) Copyright IBM Corp. 2021
+#
+# SPDX-License-Identifier: Apache-2.0
+###############################################################################
+
+set -ex
+
+# Gathers the logs
+package_logs(){
```
Comment (lmsurpre) [inline build/reindex/bin/gather-logs.sh:12]: I wonder if this one should be shared accross the various e2e tests.  Maybe introduce a `build/common` or something?

File: build/reindex/bin/integration-test.sh [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        mvn -B test -f fhir-server-test -DskipWebSocketTest=true --no-transfer-progress \
+            -DskipTests=false -Dtest="com.ibm.fhir.server.test.operation.ReindexOperationPhase1Test" | tee build/reindex/${reindex}/workarea/${reindex}-test3.log
+
+        # Update SPs
+        cp -pr ${WORKSPACE}/fhir-server-test/src/test/resources/testdata/reindex-operation/extension-search-parameters-test1.json \
+            ${WORKSPACE}/build/reindex/${reindex}/workarea/volumes/dist/config/default/extension-search-parameters.json
+
+        # Restart
+        cd build/reindex/${reindex}
+        docker-compose restart --timeout 30 fhir
```
Comment (lmsurpre) [inline build/reindex/bin/integration-test.sh:76]: I've had bad luck with `docker-compose restart fhir` locally... have you seen any problems with this?
Maybe I just need to add the timeout to mine?

File: build/reindex/db2/create-database.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
+###############################################################################
+# (C) Copyright IBM Corp. 2021
+#
+# SPDX-License-Identifier: Apache-2.0
+###############################################################################
+
+# Set the instance_memory and enable to 90 days trail license to avoid memory allocation failures.
```
Comment (lmsurpre) [inline build/reindex/db2/create-database.sh:8]:
```suggestion
# Set the instance_memory and enable 90 day trial license to avoid memory allocation failures.
```

File: fhir-examples/src/main/resources/json/ibm/basic/BasicDate.json [hunk 4]
```diff
@@ -98,7 +98,7 @@
         "_start": {
           "extension": [{
             "url": "http://hl7.org/fhir/StructureDefinition/data-absent-reason",
-            "valueString": "masked"
+            "valueCode": "masked"
```
Comment (lmsurpre) [inline fhir-examples/src/main/resources/json/ibm/basic/BasicDate.json:101]: is this fix needed for any tests, or just a general cleanup thing?   just checking since we currently point to the released fhir-examples.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase2Test.java [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+import com.ibm.fhir.model.resource.OperationOutcome;
+import com.ibm.fhir.model.resource.Parameters;
+import com.ibm.fhir.model.resource.Parameters.Parameter;
+import com.ibm.fhir.model.test.TestUtil;
+import com.ibm.fhir.path.exception.FHIRPathException;
+import com.ibm.fhir.server.test.FHIRServerTestBase;
+
+/**
+ * This integration tests the <b>fhir-operation-reindex</b> $reindex operation.
+ * Phase 2: Completes the reindex changes
```
Comment (lmsurpre) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase2Test.java:38]: Not quite sure what this means.  It looks like it performs a number of searches and confirms we get results.
What doesn't it do?

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase3Test.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import com.ibm.fhir.model.resource.OperationOutcome;
+import com.ibm.fhir.model.resource.Parameters;
+import com.ibm.fhir.model.resource.Parameters.Parameter;
+import com.ibm.fhir.model.test.TestUtil;
+import com.ibm.fhir.path.exception.FHIRPathException;
+import com.ibm.fhir.server.test.FHIRServerTestBase;
+
+/**
+ * This integration tests the <b>fhir-operation-reindex</b> $reindex operation.
+ * Phase 3: Completes the reindex changes
```
Comment (lmsurpre) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase3Test.java:37]: what does this mean?  how does it differ from phase 2?

File: docs/src/pages/guides/FHIRSearchConfiguration.md [hunk 7]
```diff
@@ -269,6 +269,51 @@ The `$reindex` operation can be invoked via an HTTP(s) POST to `[base]/$reindex`
 |----|----|-----------|
 |`tstamp`|string|Reindex any resource not previously reindexed before this timestamp. Format as a date YYYY-MM-DD or time YYYY-MM-DDTHH:MM:DDZ.|
 |`resourceCount`|integer|The maximum number of resources to reindex in this call. If this number is too large, the processing time might exceed the transaction timeout and fail.|
+|`resourceLogicalId`|string|The ResourceType or the ResourceType/Logical id for targetted reindexing|
```
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRSearchConfiguration.md:272]: Good documentation add.  A related thought on the design of this operation is that it might be more intuitive to support this capability by exposing `$reindex` at both the type and the index level.
This way, we wouldn't really need this parameter at all.

`POST [base]/$reindex` would reindex resources of any type
`POST [base]/Patient/$reindex` would reindex Patient resources
`POST [base]/Patient/123/$reindex` would reindex just Patient/123

Probably a better thought for #2195 ...

File: operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java [hunk 8]
```diff
@@ -71,7 +72,7 @@ protected Parameters doInvoke(FHIROperationContext operationContext, Class<? ext
         // Allow only POST because we're changing the state of the database
         String method = (String) operationContext.getProperty(FHIROperationContext.PROPNAME_METHOD_TYPE);
         if (!"POST".equalsIgnoreCase(method)) {
-            throw new FHIROperationException("HTTP method not supported: " + method);
+            throw FHIROperationUtil.buildExceptionWithIssue("HTTP method not supported: " + method, IssueType.NOT_SUPPORTED);
```
Comment (lmsurpre) [inline operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java:75]: It reminds me that I'd like to move this check to our operations framework and base it on the value of this element:  https://www.hl7.org/fhir/operationdefinition-definitions.html#OperationDefinition.affectsState

File: docs/src/pages/guides/FHIRSearchConfiguration.md (hunk 7 above)
Comment (prb112) [reply docs/src/pages/guides/FHIRSearchConfiguration.md:272]: I think it's an easy addition that I can add

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase3Test.java (hunk 6 above)
Comment (prb112) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase3Test.java:37]: whoops  - I'll update the comments.  It should say Phase 2 slightly differently

File: fhir-examples/src/main/resources/json/ibm/basic/BasicDate.json (hunk 4 above)
Comment (prb112) [reply fhir-examples/src/main/resources/json/ibm/basic/BasicDate.json:101]: I have to update, otherwise I have to exclude it when I post to the server, which rejects it as invalid (no-value) data-absent-reason

File: build/reindex/bin/integration-test.sh (hunk 2 above)
Comment (prb112) [reply build/reindex/bin/integration-test.sh:76]: I've had pretty good luck with this... I've just been doing the fhir server though... I've had bad luck with db/fhir at the same time

File: operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java [hunk 9]
```diff
@@ -125,12 +133,18 @@ protected Parameters doInvoke(FHIROperationContext operationContext, Class<? ext
...
+                    .severity(IssueSeverity.INFORMATION)
+                    .diagnostics(com.ibm.fhir.model.type.String.of(diag))
+                    .build());
             }
 
             OperationOutcome operationOutcome = result.build();
             checkOperationOutcome(operationOutcome);
             return FHIROperationUtil.getOutputParameters(operationOutcome);
+        } catch (java.time.format.DateTimeParseException dtpe) {
+            throw FHIROperationUtil.buildExceptionWithIssue("Invalid format for 'tstamp' value, 'yyyy-MM-dd' or iso-date format are valid", IssueType.INVALID);
```
Comment (lmsurpre) [inline operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java:147]:
```suggestion
            throw FHIROperationUtil.buildExceptionWithIssue("Invalid format for 'tstamp' value, only 'yyyy-MM-dd' or ISO 8601 dateTime format is valid", IssueType.INVALID);
```

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase2Test.java (hunk 5 above)
Comment (prb112) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase2Test.java:38]: Updated to  * <b>Phase 2</b>:
 * Pre-condition: The Phase 1 loaded Resources which were not previously found.
 * Completes the reindexes with the newly found SearchParameters.
 * These SearchParameters include the original FHIR Path Expressions important for the `ChangedExpression`.
 *
 * This test is followed by a Server Side Restart and then Phase3.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase3Test.java (hunk 6 above)
Comment (prb112) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase3Test.java:37]: `Phase 3: Completes the reindex changes after Expression Change`

File: docs/src/pages/guides/FHIRSearchConfiguration.md (hunk 7 above)
Comment (prb112) [reply docs/src/pages/guides/FHIRSearchConfiguration.md:272]: I'd like to get this in a follow on PR.

Comment (prb112) [reply docs/src/pages/guides/FHIRSearchConfiguration.md:272]: Support $reindex operation supports resourceLogicalId on Type and Instance #2503

File: build/reindex/db2/create-database.sh (hunk 3 above)
Comment (prb112) [reply build/reindex/db2/create-database.sh:8]: Fixed locally

File: operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java (hunk 8 above)
Comment (prb112) [reply operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java:75]: https://github.com/IBM/FHIR/issues/2504

File: build/reindex/bin/gather-logs.sh (hunk 1 above)
Comment (prb112) [reply build/reindex/bin/gather-logs.sh]: moved.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase1Test.java [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+                        .build();
+                Entity<Resource> entity = Entity.entity(r, FHIRMediaType.APPLICATION_FHIR_JSON);
+                Response response = getWebTarget()
+                        .path("Patient/REIN-DEX-TEST-1")
+                        .request()
+                        .put(entity, Response.class);
+                assertEquals(Response.Status.Family.familyOf(response.getStatus()), Response.Status.Family.SUCCESSFUL);
+            }
+
+            // Search and Confirm the search using the test-code1 does not return anything.
```
Comment (tbieste) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase1Test.java:86]: Maybe rephrase these "does not return anything" comments in this file to indicate that BAD_REQUEST is due to the search parameter not recognized, and not due to the search not finding any matching resources.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase2Test.java [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+        Bundle bundle = response.readEntity(Bundle.class);
+        assertFalse(bundle.getEntry().isEmpty());
+        assertEquals(bundle.getEntry().get(0).getId(), id);
+    }
+
+    @Test(groups = {"reindex"}, dependsOnMethods = {})
+    public void testReindex_String_Phase2() throws IOException, FHIRParserException, FHIRPathException {
+        if (runIt) {
+            reindex("REINDEX-STRING");
+            verifyFoundBySearch("reindex-string", "REINDEX-STRING", "");
```
Comment (tbieste) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase2Test.java:222]: Shouldn't the value be "testString" to match BasicString.json?

File: operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java [hunk 12]
```diff
@@ -101,14 +110,25 @@ protected Parameters doInvoke(FHIROperationContext operationContext, Class<? ext
...
+                            throw FHIROperationUtil.buildExceptionWithIssue("resourceLogicalId already specified using call to Operation on Type or Instance", IssueType.INVALID);
+                        }
+                        // reindex a specific resource or resourceType
                         resourceLogicalId = parameter.getValue().as(com.ibm.fhir.model.type.String.class).getValue();
+                        String rt = resourceLogicalId;
+                        if (resourceLogicalId.contains("/")) {
+                            String[] parts = resourceLogicalId.split("/");
+                            rt = parts[0];
+                        }
+                        if (!ModelSupport.isResourceType(rt)) {
```
Comment (tbieste) [inline operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java:129]: Disallow abstract types?

File: operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java [hunk 13]
```diff
@@ -71,14 +72,23 @@ protected Parameters doInvoke(FHIROperationContext operationContext, Class<? ext
         // Allow only POST because we're changing the state of the database
```
Comment (tbieste) [inline operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java:72]: Update copyright year.

File: build/reindex/derby/docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+        source: ./workarea/volumes/dist/userlib/
+        target: /opt/ol/wlp/usr/servers/defaultServer/userlib
+        read_only: true
+    command: bash -c "/opt/ol/wlp/bin/server run"
+    healthcheck:
+      start_period: 120s
+      interval: 10s
+      timeout: 10s
+      retries: 2
+      test: curl --max-time 5 -v -f -k -u 'fhiruser:change-password' 'https://localhost:9443/fhir-server/api/v4/$$healthcheck'
```
[the remark this answers is not in the document]
Comment (prb112) [reply build/reindex/derby/docker-compose.yml:21]: this is a docker limitation.  It's correct.

File: build/reindex/postgres/docker-compose.yml [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        read_only: true
+    command: bash -c "
+      java -jar /opt/ibm-fhir-server/tools/fhir-persistence-schema-*-cli.jar --db-type postgresql --prop db.host=postgres_postgres_1 --prop db.port=5432 --prop db.database=fhirdb --prop user=fhiradmin --prop password=change-password --update-schema --grant-to fhirserver &&
+      /opt/ol/wlp/bin/server run"
+    healthcheck:
+      start_period: 32s
+      interval: 30s
+      timeout: 5s
+      retries: 3
+      test: curl -k -u 'fhiruser:change-password' 'https://localhost:9443/fhir-server/api/v4/$$healthcheck'
```
[the remark this answers is not in the document]
Comment (prb112) [reply build/reindex/postgres/docker-compose.yml:59]: this is a docker limitation.  It's correct.

File: build/reindex/db2/docker-compose.yml [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        target: /opt/ol/wlp/usr/servers/defaultServer/userlib
+        read_only: true
+      - ./resources/startup.sh:/opt/ol/wlp/usr/servers/defaultServer/startup.sh
+    command: sh -c "chmod +x /opt/ol/wlp/usr/servers/defaultServer/startup.sh && /opt/ol/wlp/usr/servers/defaultServer/startup.sh"
+    healthcheck:
+      start_period: 120s
+      interval: 10s
+      timeout: 10s
+      retries: 2
+      test: curl --max-time 5 -v -f -k -u 'fhiruser:change-password' 'https://localhost:9443/fhir-server/api/v4/$$healthcheck'
```
[the remark this answers is not in the document]
Comment (prb112) [reply build/reindex/db2/docker-compose.yml:52]: this is a docker limitation.  It's correct.

File: operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java (hunk 13 above)
Comment (prb112) [reply operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java:72]: Done

File: operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java (hunk 12 above)
Comment (prb112) [reply operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java:129]: Done

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase2Test.java (hunk 11 above)
Comment (prb112) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/ReindexOperationPhase2Test.java:222]: Fixed

File: docs/src/pages/guides/FHIRSearchConfiguration.md [hunk 17]
```diff
@@ -263,12 +263,56 @@ In order to avoid this issue, inclusion criteria search parameters should not be
 ##  2 Re-index
 Reindexing is implemented as a custom operation that tells the IBM FHIR Server to read a set of resources and replace the existing search parameters with those newly extracted from the resource body.
 
-The `$reindex` operation can be invoked via an HTTP(s) POST to `[base]/$reindex`. By default, the operation will select 10 resources and re-extract their search parameters values based on the current configuration of the server. The operation supports the following parameters to control the behavior:
+The `$reindex` operation can be invoked via an HTTP(s) POST to `[base]/$reindex`, `[base]/[type]/$reindex` or , `[base]/[type]/[instance]/$reindex`. By default, the operation at the System-level or Type-level selects 10 resources and re-extract their search parameters values based on the current configuration of the server. The operation supports the following parameters to control the behavior:
```
[the remark this answers is not in the document]
Comment (prb112) [reply docs/src/pages/guides/FHIRSearchConfiguration.md:266]: Fixed

**Review by tbieste — APPROVED** [review]
LGTM (when build successful)

**Review by lmsurpre — APPROVED** [review]
LGTM
